test/jdk/java/lang/ProcessBuilder/Basic.java

```diff
@@ -2179,7 +2179,7 @@ public void run() {
 
                 thread.start();
                 latch.await();
-                Thread.sleep(10);
+                Thread.sleep(100L);     // Wait for child initialization to settle
 
```

Comment:
Hi Roger,
Shouldn't the code that follows this sleep already be enough to wait for child thread to get a lock on the stream?

```
                Thread.sleep(100L);     // Wait for child initialization to settle

                if (s instanceof BufferedInputStream) {
                    // Wait until after the s.read occurs in "thread" by
                    // checking when the input stream monitor is acquired
                    // (BufferedInputStream.read is synchronized)
                    while (!isLocked(s, 10)) {
                        Thread.sleep(100);
                    }
                }
```  

From 1st glance I would conclude that the 1st sleep is unnecessary.

Reply:
That loop is checking that the Thread (in the parent) reading from the child is in the correct state (blocked).  On Windows, it is a BufferedInputStream.

But it does not indicate anything about the state of the child process.
From the scant information from previous failures, it appears that the the creation of some thread (not a java thread) in the child has failed.
The additional time is to avoid destroying the child while it is still initializing.

Reply:
The failures happened in tiers 6 and 8. The system may be overloaded so even 100ms may not be enough for the child process to start sleeping. From the error log, the child process tried to spawn a thread (probably one of those usually started during VM bootstrap) at around 118ms

```
[0.118s][warning][os,thread] Failed to start thread - _beginthreadex failed (EACCES) for attributes: stacksize: default, flags: 
```

The test runs 4 times. Each time it checks only STDOUT or STDERR, but not both. So I think we can use the other stream to signal to the main process that the child process is ready. That would be more reliable than an arbitrary wait time.

Reply:
That complicates the test and the child quite a bit for minimal gain.

Reply:
Arbitrary time out has been a reliable source of intermittent failures.

Since we have spent a lot of time analyzing this failure, I think it's worthwhile to fix it properly, which doesn't seem that complicated. That's better than the same bug happening again a year later and a different set of people would spend hours to analyze it again.

Reply:
I'm not sure what part of the system or tests should be more robust.
If the VM can't tolerate the failure of a thread to start, then it should be a fatal VM error, not just an unexpected line on stdout (or it should be able to proceed without the extra thread).
It is a reasonable position to take that when a process is destroyed, there can be unpredictable effects.
Though in various iterations, attempts have been made to shutdown in an orderly fashion, coupled to calls to System.exit() or SIGTERM.
(It would be nice if Windows had a way to do a cleanly request process termination - a SIGTERM equivalent does not exist.)
Avoidance may be easiest but may just hide another problem.

Reply:
@tstuefe It's unlikely that _beginthreadex failed due to lack of memory. We are running on a machine with more than 50GB ram with only concurrency of 6.  Extracts from the logs:

```
$ jtreg -vmoption:-Xmx512m -concurrency:6 -vmoption:-XX:MaxRAMPercentage=4 .... open/test/jdk:jdk_lang

start = Wed Mar 10 22:54:53 GMT 2021
end = Wed Mar 10 22:56:36 GMT 2021
elapsed= 102932 0:01:42.932

----------------------------------------
[2021-03-10 22:57:07] [C:\cygwin\bin\free.exe] timeout=20000
----------------------------------------
              total used free shared buff/cache available
Mem: 50068964 19974888 30094076 0 0 30094076
Swap: 8388608 105048 8283560
----------------------------------------
[2021-03-10 22:57:07] exit code: 0 time: 33 ms
----------------------------------------
```

My theory is that `TerminateProcess` has made it impossible for the child process to spawn new threads, but somehow existing threads are still able to run a little bit and produce the log message.

Of course this is just a theory, and I cannot find any supporting docs from MS. However, if we implement the work around and make sure we don't kill the child until it has finished bootstrapping, and the bug doesn't happen anymore, then we know something more.

Reply:
Hmm. That's strange. 

One issue I don't understand with this test, shouldn't the parent not read from both streams simultaneously, because if the child is sending output to the stream the parent is not reading from the child may block on write? 

This depends on file io buffer size of course but if the child writes some larger output into e.g. stderr while the parent reads stdout, parent will wait on stdout and child will wait on the write to stderr finishing. So, either we should read both streams or redirect the stream we are not interested in to paren'ts stdout/stderr where hopefully the testrunner itself would read it.

I may misunderstand the test completely; if yes, sorry for the confusion!

Thomas

Reply:
The child does no (zero) writes to either stream.  It is invoked only to sleep until it is destroyed.
The purpose of the test is to verify the exception that is thrown when the other end(child) of the pipe is closed (because the process has been forcibly terminated).

Reply:
Hmm, maybe the child can create a file to indicate that bootstrap has finished.

Reply:
> The child does no (zero) writes to either stream. It is invoked only to sleep until it is destroyed.
> The purpose of the test is to verify the exception that is thrown when the other end(child) of the pipe is closed (because the process has been forcibly terminated).

I meant as a consequence of an error. E.g. a crash or a native OOM could cause the child to write an error report to stderr.

Reply:
> Hmm, maybe the child can create a file to indicate that bootstrap has finished.

Why does the child even have to be a java VM?

Reply:
Its a Java Child for consistency across tests and across OS's.  
The JavaChild executes a number of specialized commands to consume or provide data to the parent.
Piecing that together on different OS's would add more variables to the tests.
In this particular case, assuming it's well understood could probably be handled by some native program
as long its available on all platforms consistently.

Reply:
> Its a Java Child for consistency across tests and across OS's.
> The JavaChild executes a number of specialized commands to consume or provide data to the parent.
> Piecing that together on different OS's would add more variables to the tests.
> In this particular case, assuming it's well understood could probably be handled by some native program
> as long its available on all platforms consistently.

Thanks Roger for explaining. Sorry for the questions. We are bugged regularly by similar intermittent errors and I am curious and interested in getting tests stable.

I'm still thinking reading both stdout and stderr simultaneously would be more defensive in case the child dies of unnatural circumstances, or prints out unexpected output. The VM does this sometimes (eg unconditonal logging, or crashing), and we have seen blockages like these when childs waited on unflushed write buffers. Since you have all parts already there - a thread to drain the childs output - why not just spawn two of those instead of one and make sure both are closed as expected when the child is gone. That way you also cut down on execution time.

Reply:
The test expects there to be zero output from the child (and it doesn't matter what state the child is in).
Can the logging from the VM be disabled or re-directed?

Reply:
> That complicates the test and the child quite a bit for minimal gain.

Suppose that the child process writes 1 byte to STDOUT and 1 byte to STDERR. In the monitoring thread in the parent process, you could read 1 byte from either STDOUT or STDERR, then count-down the latch and then block in next read. The main thread could then just wait for the latch, wait for the monitoring thread to re-enter the synchronized read method and then destroy the process... It doesn't sound too complicated. WDYT?

Reply:
Redirecting the stream not being tested to INHERIT may be useful.
But the failing case is when the vm output is being directed to the stream being tested.
The command line arguments are:
```
    Arrays.asList(javaExe,
        "-XX:+DisplayVMOutputToStderr",
        "-classpath", absolutifyPath(classpath),
        "Basic$JavaChild");```
A combination of redirecting to inherited and sending the VM output to the same (untested) stream seems like a good option.

Reply:
@plevart, I prototyped using that technique but was concerned that it might change the test case to be different in some subtle way from the original. It adds a new subcommand to the JavaChild to add the specific behavior.
Not particularly complex but more code and dependencies to explain.